apps/recnet-api/src/modules/rec/rec.service.ts

```diff
@@ -0,0 +1,248 @@
+import UserRepository from "@recnet-api/database/repository/user.repository";
+import { getOffset } from "@recnet-api/utils";
+import { RecnetError } from "@recnet-api/utils/error/recnet.error";
+import { ErrorCode } from "@recnet-api/utils/error/recnet.error.const";
+
+import { getNextCutOff } from "@recnet/recnet-date-fns";
+
+import { CreateArticleDto } from "./dto/create.rec.dto";
+import { UpdateArticleDto } from "./dto/update.rec.dto";
+import { Rec } from "./entities/rec.entity";
+import {
+  CreateRecResponse,
+  GetFeedsResponse,
+  GetRecsResponse,
+  GetUpcomingRecResponse,
+  UpdateRecResponse,
+} from "./rec.response";
+
+@Injectable()
+export class RecService {
+  constructor(
+    @Inject(RecRepository)
+    private readonly recRepository: RecRepository,
+    @Inject(UserRepository)
+    private readonly userRepository: UserRepository
+  ) {}
+
+  public async getRecs(
+    page: number,
+    pageSize: number,
+    userId: string
+  ): Promise<GetRecsResponse> {
+    const recCount = await this.recRepository.countRecs({
+      userId: userId,
+    });
+    const dbRecs = await this.recRepository.findRecs(page, pageSize, {
+      userId: userId,
+    });
+    const recs = this.getRecsFromDbRecs(dbRecs);
+
+    return {
+      hasNext: recs.length + getOffset(page, pageSize) < recCount,
+      recs: recs,
+    };
+  }
+
+  public async getFeeds(
+    page: number,
+    pageSize: number,
+    cutoff: number,
+    userId: string
+  ): Promise<GetFeedsResponse> {
+    const user = await this.userRepository.getUser(userId);
+    const followings = user.following.map((following) => following.followingId);
+    const where: Prisma.RecommendationWhereInput = {
+      userId: {
+        in: followings,
+      },
+      cutoff: new Date(cutoff),
+    };
+    const recCount = await this.recRepository.countRecs(where);
+    const dbRecs = await this.recRepository.findRecs(page, pageSize, where);
+    const recs = this.getRecsFromDbRecs(dbRecs);
+
+    return {
+      hasNext: recs.length + getOffset(page, pageSize) < recCount,
+      recs: recs,
+    };
+  }
+
+  public async getUpcomingRec(userId: string): Promise<GetUpcomingRecResponse> {
+    const dbRec = await this.recRepository.findUpcomingRec(userId);
+    if (!dbRec) {
+      return {
+        rec: null,
+      };
+    }
+    return {
+      rec: this.getRecFromDbRec(dbRec),
+    };
+  }
+
+  public async addRec(
+    articleId: string | null,
+    article: CreateArticleDto | null,
+    description: string,
+    userId: string
+  ): Promise<CreateRecResponse> {
+    // article and articleId cannot be null at the same time
+    // check libs/recnet-api-model/src/lib/api/rec.ts
+    if (!article && !articleId) {
+      // which code should be returned?
+      throw new RecnetError(
+        ErrorCode.INTERNAL_SERVER_ERROR,
+        HttpStatus.BAD_REQUEST,
+        "Article and articleId cannot be null at the same time"
+      );
```

Comment:
Here I don't think we should throw `ErrorCode.INTERNAL_SERVER_ERROR` since in such case the error is caused by the one who's calling the API with wrong payload, not the server.

Reply:
I just put an INTERNAL_SERVER_ERROR for now since there's only `INTERNAL_SERVER_ERROR` and `ZOD_VALIDATION_ERROR` in that enum. I think we should extend it and create another error node. But, I think before doing that, we should have a discussion. So I just leave it there for now.

Reply:
Do you think we should create another custom error code like `BAD_REQUEST` maybe map to 1002 and use it here?

Reply:
That said, actually this block will never be triggered since if two fields are `null` at the same time, it will be catch be the `ZodValidationPipe` since the req body schema looks like this

```ts
export const recFormSubmissionSchema = z.intersection(
  z.union([
    z.object({
      articleId: z.string(),
      article: z.null(),
    }),
    z.object({
      articleId: z.null(),
      article: articleSchema.omit({ id: true, isVerified: true }),
    }),
  ]),
  z.object({
    description: z.string(),
  })
);
export type RecFormSubmission = z.infer<typeof recFormSubmissionSchema>;
```

If two fields are null at the same time, it will throw error earlier. But since DTO can't produce such fine-grained type. This block and the if statement is just used for narrowing down the type.